packages/bsp/common/usr/lib/armbian/armbian-resize-filesystem

```diff
@@ -13,208 +13,287 @@
-		if [[ $capacity -ge 5 ]]; then
-			local secondpartition=$(( 32768 * $(parted $rootdevicepath unit s print -sm | grep "^$rootdevicepath" | awk -F":" "{printf (\"%0d\", ( \$2 * 99 / 3276800))}") -1 ))
-			if [[ $secondpartition -lt $partend ]]; then
-				unset secondpartition
-			fi
+
+		# if remain more than 1GiB then create another partition
+		if [[ $(( $sectorsize * ($capacity - $lastsector) )) -ge $(( 1 * 1024 * 1024 * 1024 )) ]]; then
+			local newpartition=$(( $capacity * 99 / 100 ))
 		fi
 	else
 		# check device capacity. If 4GiB or below do not use whole card but leave a 5% spare area
 		# to help older cards with wear leveling and garbage collection. In case this reduced card
 		# capacity is less than the actual image capacity this is a clear sign that someone wants
 		# to use Armbian on a card of inappropriate size so he gets what he deserves (at least he
 		# should know what he's doing)
-		if [[ $capacity -lt 5 ]]; then # 4 GiB or less
-			local lastsector=$(( 32768 * $(parted $rootdevicepath unit s print -sm | grep "^$rootdevicepath" | awk -F":" "{printf (\"%0d\", ( \$2 * 95 / 3276800))}") -1 ))
-			if [[ $lastsector -lt $partend ]]; then
-				unset lastsector
-			else
-				ResizeLog="4GiB or smaller media - leaving 5% spare area"
-			fi
-		elif [[ $capacity -lt 9 ]]; then # 8 GiB or less
+		if [[ $(( $sectorsize * $capacity )) -lt $(( 4 * 1024 * 1024 * 1024 )) ]]; then # 4 GiB or less
+			# Leave 5 percent unpartitioned
+			local lastsector=$(( $capacity * 95 / 100 ))
+			echo "4GiB or smaller media - leaving 5% spare area"
+		elif [[ $(( $sectorsize * $capacity )) -lt $(( 8 * 1024 * 1024 * 1024 )) ]]; then # 8 GiB or less
 			# Leave 2 percent unpartitioned
-			local lastsector=$(( 32768 * $(parted $rootdevicepath unit s print -sm | grep "^$rootdevicepath" | awk -F":" "{printf (\"%0d\", ( \$2 * 98 / 3276800))}") -1 ))
-			if [[ $lastsector -lt $partend ]]; then
-				unset lastsector
-			else
-				ResizeLog="8GiB or smaller media - leaving 2% spare area"
-			fi
+			local lastsector=$(( $capacity * 98 / 100 ))
+			echo "8GiB or smaller media - leaving 2% spare area"
 		else
 			# Leave 1 percent unpartitioned
-			local lastsector=$(( 32768 * $(parted $rootdevicepath unit s print -sm | grep "^$rootdevicepath" | awk -F":" "{printf (\"%0d\", ( \$2 * 99 / 3276800))}") -1 ))
-			if [[ $lastsector -lt $partend ]]; then
-				unset lastsector
-			else
-				ResizeLog="Leaving 1% spare area"
-			fi
+			local lastsector=$(( $capacity * 99 / 100 ))
+			echo "Leaving 1% spare area"
 		fi
 	fi
 
+	# use 16MiB to align partitions since this is the erase block size of more recent SD cards
+	local lastsector=$(( $lastsector - ($lastsector % (16 * 1024 * 1024 / $sectorsize)) ))
+	[[ -n "$newpartition" ]] && local newpartition=$(( $newpartition - ($newpartition % (16 * 1024 * 1024 / $sectorsize)) ))
+
+	if [[ $lastsector -lt $partend ]]; then
+		echo -e "\nWe can't shrink the partition from ${partend}s to ${lastsector}s" >&2
```

Comment:
As I can see this tool just can't shrink. Then maybe better say it clear?
Like (sorry for my chilldish language) "$partend is larger than $lastsector, that means you trying to shrink filesystem. Sorry, this tool designed for stretching only"

It's only idea.

Reply:
my english is really bad. so the message may incorrent.
I change it